src/browser/dom/range.zig

```diff
@@ -0,0 +1,122 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+const std = @import("std");
+
+const log = @import("../../log.zig");
+const parser = @import("../netsurf.zig");
+
+const Page = @import("../page.zig").Page;
+
+pub const Interfaces = .{
+    AbstractRange,
+    Range,
+};
+
+pub const AbstractRange = struct {
+    collapsed: bool,
+    end_container: *parser.Node,
+    end_offset: i32,
+    start_container: *parser.Node,
+    start_offset: i32,
+
+    pub fn updateCollapsed(self: *AbstractRange) void {
+        // TODO: Eventually, compare properly.
+        self.collapsed = false;
+    }
+
+    pub fn get_collapsed(self: *const AbstractRange) bool {
+        return self.collapsed;
+    }
+
+    pub fn get_endContainer(self: *const AbstractRange) *parser.Node {
+        return self.end_container;
+    }
+
+    pub fn get_endOffset(self: *const AbstractRange) i32 {
+        return self.end_offset;
+    }
+
+    pub fn get_startContainer(self: *const AbstractRange) *parser.Node {
+        return self.start_container;
+    }
+
+    pub fn get_startOffset(self: *const AbstractRange) i32 {
+        return self.start_offset;
+    }
+};
+
+pub const Range = struct {
+    pub const prototype = *AbstractRange;
+
+    proto: AbstractRange,
+
+    pub fn constructor() Range {
+        const proto: AbstractRange = .{
+            .collapsed = true,
+            .end_container = undefined,
+            .end_offset = 0,
+            .start_container = undefined,
+            .start_offset = 0,
+        };
+
+        return .{ .proto = proto };
+    }
+
+    pub fn _setStart(self: *Range, node: *parser.Node, offset: i32) void {
+        self.proto.start_container = node;
+        self.proto.start_offset = offset;
+        self.proto.updateCollapsed();
+    }
+
+    pub fn _setEnd(self: *Range, node: *parser.Node, offset: i32) void {
+        self.proto.end_container = node;
+        self.proto.end_offset = offset;
+        self.proto.updateCollapsed();
+    }
+
+    pub fn _createContextualFragment(_: *Range, fragment: []const u8, page: *Page) !*parser.DocumentFragment {
+        const document_html = page.window.document;
+        const document = parser.documentHTMLToDocument(document_html);
+        const doc_frag = try parser.documentParseFragmentFromStr(document, fragment);
+        return doc_frag;
+    }
+};
+
+const testing = @import("../../testing.zig");
+test "Browser.Range" {
+    var runner = try testing.jsRunner(testing.tracking_allocator, .{});
+    defer runner.deinit();
+
+    try runner.testCases(&.{
+        // Test Range constructor
+        .{ "let range = new Range()", "undefined" },
```

Comment:
This is fine, but just FYI, if you don't care about the output, you can pass `null` as the 2nd parameter:

```zig
.{ "let range = new Range()", null },
```

But maybe in this case you _do_ care to check that its `undefined`